llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -329,6 +334,51 @@ static void emitDirectiveRelocJalr(const MachineInstr &MI,
   }
 }
 
+// NOTE: This is being used to control optimization of BALC in the linker.
+// TODO: Introduce a new relocation that will serve for this purpose only.
+static bool emitDirectiveRelocBalcStubs(const MachineInstr &MI,
+                                        MCContext &OutContext,
+                                        TargetMachine &TM,
+                                        MCStreamer &OutStreamer) {
+  assert(EmitNanoMipsBalcStubsReloc);
+  if (MI.getMF() && MI.getMF()->getFunction().hasOptSize()) {
+    MCSymbol *OffsetLabel = OutContext.createTempSymbol();
+    const MCExpr *OffsetExpr = MCSymbolRefExpr::create(OffsetLabel, OutContext);
+    // This symbol will be used to indicate that we do want balc stubs
+    // optimization. The '\2' ensures that no other symbol will get the same
+    // name as this.
+    MCSymbol *DummySymbol =
+        OutContext.getOrCreateSymbol("__reloc_balc_stub_\\2");
+    DummySymbol->setExternal(true);
+    OutStreamer.emitSymbolAttribute(DummySymbol, MCSA_Global);
+
+    if (!MI.getOperand(0).isGlobal())
+      return false;
+
+    // TODO: This is a workaround for external assembler that does not allow
+    // multiple relocations on the BALC instructions.
+    MCSymbol *CallTargetSymbol =
+        OutContext.getOrCreateSymbol(MI.getOperand(0).getGlobal()->getName());
+    const MCExpr *CallTargetExpr =
+        MCSymbolRefExpr::create(CallTargetSymbol, OutContext);
+    OutStreamer.emitRelocDirective(*OffsetExpr, "R_NANOMIPS_PC25_S1",
+                                   CallTargetExpr, SMLoc(),
+                                   *TM.getMCSubtargetInfo());
+    const MCExpr *DummyExpr = MCSymbolRefExpr::create(DummySymbol, OutContext);
+    OutStreamer.emitRelocDirective(*OffsetExpr, "R_NANOMIPS_NONE", DummyExpr,
```

Comment:
See binutils [pr#2](https://github.com/MediaTek-Labs/binutils-gdb/pull/2) for R_NANOMIPS_NOTRAMP relocation